r/R/dplyr-funcs-datetime.R

```diff
@@ -147,5 +147,7 @@ register_bindings_datetime <- function() {
   register_binding("pm", function(x) {
     !call_binding("am", x)
   })
-
+  register_binding("date", function(x) {
+    x$cast(date32())
+  })
```

Comment:
Similar to https://github.com/apache/arrow/pull/12431#discussion_r810508212 are the errors that we get if `x` here is something that would be workable for our users?

Separately, what happens if you pass an R object to this instead of a column in an arrow table? Does that work ok?

Reply:
Q1: I have added some unit tests for _unsupported_ types. And discovered that, unlike {lubridate}, you can cast a `date` from an `integer` in {arrow}. It looks like the integer is interpreted as number of days since epoch. 
``` r
suppressPackageStartupMessages(library(arrow))
call_binding("date", Scalar$create(32L))
#> Scalar
#> 1970-02-02

lubridate::date(32L)
#> Error in as.POSIXlt.numeric(x, tz = tz(x)): 'origin' must be supplied
```

<sup>Created on 2022-02-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Q2: Not sure what we want to achieve by calling a binding (in this case `"date"`) on an R object. Do we want to support this (should all bindings support R objects as inputs?) or do we just want to make sure the error message is meaningful?

Reply:
> Not sure what we want to achieve by calling a binding (in this case "date") on an R object

I may be misinterpreting this, but I think maybe what Jon is getting at is whether you could do something like this in an arrow-dplyr chain:

```
library(dplyr)
library(lubridate)

arbitrary_date_object <- ymd_hms("2012-03-26 23:12:13") + days(1:32)

mtcars %>%
  mutate(x = date(arbitrary_date_object))

```

Reply:
Thanks, @thisisnic for that example. I think that sort of use of R objects is not supported at all in an _arrow_ pipeline (see my _reprex_ below). So neither the binding for `date()`, nor any of the other lubridate ones would work in that context. 

``` r
suppressPackageStartupMessages(library(arrow))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(lubridate))

arbitrary_date_object <- ymd_hms("2012-03-26 23:12:13") + days(1:32)
# dplyr pipeline
mtcars %>% 
  mutate(x = arbitrary_date_object) %>% 
  head(4)
#>                 mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Mazda RX4      21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
#> Mazda RX4 Wag  21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
#> Datsun 710     22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
#> Hornet 4 Drive 21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
#>                                  x
#> Mazda RX4      2012-03-27 23:12:13
#> Mazda RX4 Wag  2012-03-28 23:12:13
#> Datsun 710     2012-03-29 23:12:13
#> Hornet 4 Drive 2012-03-30 23:12:13

# arrow pipeline
mtcars %>% 
  arrow_table() %>% 
  mutate(x = arbitrary_date_object) %>% 
  collect() %>% 
  head(4)
#> Warning: In x = arbitrary_date_object, only values of size one are recycled;
#> pulling data into R
#>    mpg cyl disp  hp drat    wt  qsec vs am gear carb                   x
#> 1 21.0   6  160 110 3.90 2.620 16.46  0  1    4    4 2012-03-27 23:12:13
#> 2 21.0   6  160 110 3.90 2.875 17.02  0  1    4    4 2012-03-28 23:12:13
#> 3 22.8   4  108  93 3.85 2.320 18.61  1  1    4    1 2012-03-29 23:12:13
#> 4 21.4   6  258 110 3.08 3.215 19.44  1  0    3    1 2012-03-30 23:12:13
```

<sup>Created on 2022-02-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Take a look at that warning/error message though: recycling of length one _does_ work. 

If you want some more context on that I would recommend looking at the PR / jiras related to the error message there "only values of size one are recycled" (or possibly even comments around that code. That's what I'd do in this scenario, at least!) and that should have some more context about what/why/how that works.

Reply:
Recycling a vector of length one works. 
``` r
suppressPackageStartupMessages(library(arrow))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(lubridate))

arbitrary_date_object <- ymd_hms("2012-03-26 23:12:13")
# dplyr pipeline
mtcars %>% 
  mutate(x = arbitrary_date_object) %>% 
  head(4)
#>                 mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Mazda RX4      21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
#> Mazda RX4 Wag  21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
#> Datsun 710     22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
#> Hornet 4 Drive 21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
#>                                  x
#> Mazda RX4      2012-03-26 23:12:13
#> Mazda RX4 Wag  2012-03-26 23:12:13
#> Datsun 710     2012-03-26 23:12:13
#> Hornet 4 Drive 2012-03-26 23:12:13

# arrow pipeline
mtcars %>% 
  arrow_table() %>% 
  mutate(x = arbitrary_date_object) %>% 
  collect() %>% 
  head(4)
#>    mpg cyl disp  hp drat    wt  qsec vs am gear carb                   x
#> 1 21.0   6  160 110 3.90 2.620 16.46  0  1    4    4 2012-03-26 23:12:13
#> 2 21.0   6  160 110 3.90 2.875 17.02  0  1    4    4 2012-03-26 23:12:13
#> 3 22.8   4  108  93 3.85 2.320 18.61  1  1    4    1 2012-03-26 23:12:13
#> 4 21.4   6  258 110 3.08 3.215 19.44  1  0    3    1 2012-03-26 23:12:13
```

<sup>Created on 2022-02-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>